### ISSUE TITLE
RFE: migrate StateDirectory= and friends back from DynamicUser=1 subdir to regular one if needed

### ISSUE THREAD

**Issue by kouros17** [systemd/systemd#12131]
**systemd version the issue has been seen with**
> 241.1023-1

**Used distribution**
 > Manjaro

From Journalctl I get:

```
manjaro systemd[694]: systemd-timesyncd.service: Failed at step STATE_DIRECTORY spawning /usr/lib/systemd/systemd-timesyncd: Not a directory
manjaro systemd[694]: systemd-timesyncd.service: Failed to set up special execution directory in /var/lib: Not a directory
manjaro systemd[1]: Failed to start Network Time Synchronization.
manjaro systemd[691]: systemd-timesyncd.service: Failed at step STATE_DIRECTORY spawning /usr/lib/systemd/systemd-timesyncd: Not a directory
manjaro systemd[691]: systemd-timesyncd.service: Failed to set up special execution directory in /var/lib: Not a directory
manjaro systemd[1]: Failed to start Network Time Synchronization.
manjaro systemd[688]: systemd-timesyncd.service: Failed at step STATE_DIRECTORY spawning /usr/lib/systemd/systemd-timesyncd: Not a directory
manjaro systemd[688]: systemd-timesyncd.service: Failed to set up special execution directory in /var/lib: Not a directory
manjaro systemd[1]: Failed to start Network Time Synchronization.
manjaro systemd[685]: systemd-timesyncd.service: Failed at step STATE_DIRECTORY spawning /usr/lib/systemd/systemd-timesyncd: Not a directory
manjaro systemd[685]: systemd-timesyncd.service: Failed to set up special execution directory in /var/lib: Not a directory
manjaro systemd[1]: Failed to start Network Time Synchronization.
manjaro systemd[681]: systemd-timesyncd.service: Failed at step STATE_DIRECTORY spawning /usr/lib/systemd/systemd-timesyncd: Not a directory
manjaro systemd[681]: systemd-timesyncd.service: Failed to set up special execution directory in /var/lib: Not a directory
manjaro systemd-fsck[651]: fsck failed with exit status 16.
```

**Comment by poettering** [comment]
this suggests your /var/lib is read-only or so?

**Comment by kouros17** [comment]
ls -l /var/lib
```
σύνολο 144
drwxr-xr-x  4 root    root    4096 Οκτ  13 11:50 AccountsService
drwxr-xr-x  2 root    root    4096 Απρ   9  2018 alsa
drwxr-xr-x  2 root    root    4096 Αυγ  14  2018 arpd
drwxr-xr-x  2 root    root    4096 Σεπ  24  2018 boltd
drwxr-xr-x  4 colord  colord  4096 Οκτ  26 14:21 colord
drwxr-xr-x  2 root    root    4096 Μαρ  17 19:00 dbus
drwxr-xr-x  2 root    root    4096 Μαρ  12  2018 dhclient
drwxr-xr-x  2 root    root    4096 Αυγ  23  2018 dhcpcd
drwxrwxrwt  2 root    root    4096 Ιουν  4  2018 ex
drwx------  2 root    root    4096 Σεπ   2  2018 fprint
drwxrwx--T 13 gdm     gdm     4096 Μαρ  25 23:39 gdm
drwxr-xr-x  2 geoclue geoclue 4096 Μαρ  17 17:20 geoclue
drwxr-xr-x  4 root    root    4096 Οκτ  13 09:40 gssproxy
drwxr-xr-x  2 root    cups    4096 Νοε   4 10:48 hp
drwxr-xr-x  2 root    root    4096 Οκτ  13 09:39 krb5kdc
-rw-r--r--  1 root    root     263 Μαρ  28 00:00 logrotate.status
drwx------  2 root    root    4096 Οκτ  26 14:21 machines
drwxr-xr-x  6 root    root    4096 Οκτ  13 09:40 mhwd
drwxr-xr-x  2 root    root    4096 Σεπ  30 10:25 misc
drwx------  2 root    root    4096 Μαρ  28 09:48 NetworkManager
drwxr-xr-x  6 root    root    4096 Οκτ  13 09:40 nfs
drwxr-xr-x  2 ntp     root    4096 Μαρ  17 17:19 ntp
drwxr-xr-x  2 root    root    4096 Μαρ  17 17:20 os-prober
drwxr-xr-x  4 root    root    4096 Μαρ  28 10:00 pacman
drwxr-xr-x  2 root    root    4096 Μαρ  26 01:04 pacman-mirrors
drwxr-xr-x  2 root    root    4096 Μαρ  18 02:51 papirus-folders
drwx------  2 root    root    4096 Οκτ  26 14:21 portables
drwx------  3 root    root    4096 Οκτ  26 22:31 private
drwx------  2 rpc     rpc     4096 Οκτ  26 14:21 rpcbind
drwxr-xr-x  5 root    root    4096 Οκτ  26 22:31 systemd
drwxr-xr-x  2 root    root    4096 Οκτ  27 14:29 tlp
drwx------  2 root    root    4096 Σεπ  29 09:55 udisks2
drwxr-xr-x  2 root    root    4096 Σεπ   2  2018 upower
drwxr-xr-x  2 root    root    4096 Ιαν   5  2018 usb_modeswitch
drwxr-xr-x  2 root    root    4096 Μαρ  17 17:18 xkb
```

**Comment by poettering** [comment]
is it mounted read-only or so?

**Comment by kouros17** [comment]
How I can check this?

**Comment by kouros17** [comment]
Finally managed to solve the issue with following steps:

1) execute `rm /var/lib/systemd/timesync` to remove the old symlink
2) restart systemd-timesyncd

New directory is successfully created by systemd-timesyncd and the service starts normally

**Comment by yuwata** [comment]
C.f. 787a133f2d934d9cb01a50ee86ea5c51926c6ef8.

**Comment by floppym** [comment]
So what's the deal on this issue?
Is the sysadmin supposed to remove some file?
Or should I do this in the package upgrade post-install script?

**Comment by keszybz** [comment]
@poettering suggested adding code to move the state dirs back from private/. We currently have the code to move them one way, but not the other.

**Comment by FichteFoll** [comment]
I started the systemd-timesyncd service manually once and that made the problem go away without deleting `/var/lib/systemd/timesync`.

**Comment by poettering** [comment]
Fix waiting in #12869

**Comment by mbiebl** [comment]
I tried this with v244, but it unfortunately didn't work.
Switching a service back from DynamicUser=true to false does not automatically fix up the state directory and the start of the service fails.

**Comment by mbiebl** [comment]
```
root@pluto:~# systemctl --version
systemd 244 (244.3-1)
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid
root@pluto:~# systemctl stop systemd-timesyncd
root@pluto:~# rm -rf /var/lib/systemd/timesync/ /var/lib/private/systemd/timesync
root@pluto:~# systemctl edit --full systemd-timesyncd 
<Add DynamicUser=yes>
root@pluto:~# systemctl start systemd-timesyncd
root@pluto:~# systemctl show systemd-timesyncd.service -p DynamicUser 
DynamicUser=yes
root@pluto:~# ls -al /var/lib/systemd/timesync
lrwxrwxrwx 1 root root 27 Mär  6 04:23 /var/lib/systemd/timesync -> ../private/systemd/timesync
root@pluto:~# systemctl edit --full systemd-timesyncd
<Remove DynamicUser=yes again>
root@pluto:~# systemctl restart systemd-timesyncd
Job for systemd-timesyncd.service failed because the control process exited with error code.
See "systemctl status systemd-timesyncd.service" and "journalctl -xe" for details.
root@pluto:~# systemctl status systemd-timesyncd
● systemd-timesyncd.service - Network Time Synchronization
     Loaded: loaded (/etc/systemd/system/systemd-timesyncd.service; disabled; vendor preset: enabled)
    Drop-In: /lib/systemd/system/systemd-timesyncd.service.d
             └─disable-with-time-daemon.conf
     Active: failed (Result: exit-code) since Fri 2020-03-06 04:25:24 CET; 7s ago
       Docs: man:systemd-timesyncd.service(8)
    Process: 53523 ExecStart=/lib/systemd/systemd-timesyncd (code=exited, status=238/STATE_DIRECTORY)
   Main PID: 53523 (code=exited, status=238/STATE_DIRECTORY)

Mär 06 04:25:24 pluto systemd[1]: systemd-timesyncd.service: Scheduled restart job, restart counter is a>
Mär 06 04:25:24 pluto systemd[1]: Stopped Network Time Synchronization.
Mär 06 04:25:24 pluto systemd[1]: systemd-timesyncd.service: Start request repeated too quickly.
Mär 06 04:25:24 pluto systemd[1]: systemd-timesyncd.service: Failed with result 'exit-code'.
Mär 06 04:25:24 pluto systemd[1]: Failed to start Network Time Synchronization.
root@pluto:~# ls -al /var/lib/private/systemd/
insgesamt 12
drwxr-xr-x 3 root             root             4096 Mär  6 04:23 .
drwx------ 3 root             root             4096 Jul 23  2018 ..
drwxr-xr-x 2 systemd-timesync systemd-timesync 4096 Mär  6 04:23 timesync
```

**Comment by mbiebl** [comment]
I'm not sure if this worked with v243 and is a regression in v244 or never properly worked to begin with.

**Comment by yuwata** [comment]
Fix is waiting in #15033.
